File: kimdaeyeong/src/baekjoon/Boj21610.java
package baekjoon;

import java.util.*;
import java.io.*;

/**
 * 백준 21610
 * 마법사 상어와 비바라기
 * 골드5
 * https://www.acmicpc.net/problem/21610
 */
public class Boj21610 {

    static int n; // size of array
    static int m; // number of move
    static int[][] a; // amount of water
    static int d; // diraction
    static int s; // move distance
    static boolean[][] cloud; // current cloud location

    static int[] dx = {0, -1, -1, -1, 0, 1, 1, 1};
    static int[] dy = {-1, -1, 0, 1, 1, 1, 0, -1};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        a = new int[n][n];
        cloud = new boolean[n][n];

        for(int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for(int j = 0; j < n; j++) {
                a[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        cloud[n - 2][0] = true;
        cloud[n - 2][1] = true;
        cloud[n - 1][0] = true;
        cloud[n - 1][1] = true;

        for(int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            d = Integer.parseInt(st.nextToken());
            s = Integer.parseInt(st.nextToken());

            moveCloud(d, s);

            makeCloud();
        }

        int sum = 0;

        for(int i = 0; i < n; i++) {
            for(int j = 0; j < n; j++) {
                sum += a[i][j];
            }
        }

        System.out.println(sum);
    }

    public static void moveCloud(int d, int s) {
        boolean[][] temp = new boolean[n][n];

        for(int i = 0; i < n; i++) {
            for(int j = 0; j < n; j++) {
                if(!cloud[i][j])
                    continue;

                int nextI = i + dx[d - 1] * s;
                int nextJ = j + dy[d - 1] * s;

                if(nextI > 0)
                    nextI = nextI % n;
                if(nextJ > 0)
                    nextJ = nextJ % n;
                if(nextI < 0)
                    nextI = n - (nextI * -1 % n);
                if(nextJ < 0)
                    nextJ = n - (nextJ * -1 % n);

                if(nextI == n)
                    nextI %= n;
                if(nextJ == n)
                    nextJ %= n;

                temp[nextI][nextJ] = true;
                a[nextI][nextJ] += 1;
            }
        }

        for(int i = 0; i < n; i++)
            cloud[i] = temp[i].clone();

        for(int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (!cloud[i][j])
                    continue;

                a[i][j] += waterCopy(i, j);
            }
        }
    }

    public static int waterCopy(int x, int y) {
        int cnt = 0;

        for(int i = 1; i < 8; i += 2) {
            int nextX = x + dx[i];
            int nextY = y + dy[i];

            if(nextX < 0 || nextX >= n | nextY < 0 || nextY >= n)
                continue;

            if(a[nextX][nextY] > 0)
                cnt++;
        }

        return cnt;
    }

    public static void makeCloud() {
        boolean[][] temp = new boolean[n][n];

        for(int i = 0; i < n; i++) {
            for(int j = 0; j < n; j++) {
                if(!cloud[i][j] && a[i][j] > 1) {
                    a[i][j] -= 2;
                    temp[i][j] = true;
                }
            }
        }

        for(int i = 0; i < n; i++)
            cloud[i] = temp[i].clone();
    }
}

File: limsubin/baekjoon/boj21610.py
# BOJ 21610: 마법사 상어와 비바라기
# https://www.acmicpc.net/problem/21610

import sys

def input():
    return sys.stdin.readline().strip()

def move(d, s):
    for i in range(len(cloud)):
        x, y = cloud[i]
        mx = adjust_range(x + dx[d] * s)
        my = adjust_range(y + dy[d] * s)

        cloud[i] = [mx, my]
        visited[mx][my] = True

def adjust_range(x):
    while x < 0:
        x += n
    while x >= n:
        x -= n
    return x

def rain():
    for x, y in cloud:
        board[x][y] += 1

def copybug():
    for x, y in cloud:
        cnt = 0
        for i in range(1, 8, 2):
            mx = x + dx[i]
            my = y + dy[i]

            if mx < 0 or mx >= n or my < 0 or my >= n:
                continue

            if board[mx][my] > 0:
                cnt += 1
        board[x][y] += cnt

def make_cloud():
    temp = []
    for i in range(n):
        for j in range(n):
            # 구름이 사라진 칸 제외
            if visited[i][j]:
                continue
            if board[i][j] >= 2:
                board[i][j] -= 2
                temp.append([i, j])
    return temp

n, m = map(int, input().split())
board = [list(map(int, input().split())) for _ in range(n)]

# 구름 정보
cloud = [[i, j] for i in range(n-2, n) for j in range(2)]
visited = [[False] * n for _ in range(n)]

# 이동 방향
dx = [0, -1, -1, -1, 0, 1, 1, 1]
dy = [-1, -1, 0, 1, 1, 1, 0, -1]

for _ in range(m):
    d, s = map(int, input().split())

    # 모든 구름이 d 방향으로 s칸 이동
    move(d-1, s)

    # 비 -> 물의 양 증가
    rain()

    # 물복사버그 마법 -> 대각선 바구니 개수만큼 물의 양 증가
    copybug()

    # 구름이 사라진 칸을 제외하고 물의 양이 2 이상인 모든 칸에 구름이 생긴다.
    cloud = make_cloud()
    visited = [[False] * n for _ in range(n)]

# 바구니에 들어있는 물의 양의 합
answer = 0
for i in range(n):
    answer += sum(board[i])

print(answer)